packages/rich-text/src/SyncEditorChanges.tsx

```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
+
+import * as Contentful from '@contentful/rich-text-types';
+import { usePlateActions } from '@udecode/plate-core';
+import equal from 'fast-deep-equal';
+
+import { createOnChangeCallback } from './helpers/callbacks';
+import { usePlateSelectors } from './internal/hooks';
+import { setEditorValue } from './internal/transforms';
+import { Value } from './internal/types';
+
+/**
+ * A hook responsible for keeping the editor state in sync with incoming
+ * changes (aka. external updates
+ */
+const useAcceptIncomingChanges = (incomingValue?: Value) => {
+  const editor = usePlateSelectors().editor();
+
+  // Cache latest editor value to avoid unnecessary updates
+  const lastIncomingValue = React.useRef(incomingValue);
+
+  React.useEffect(() => {
+    if (equal(lastIncomingValue.current, incomingValue)) {
+      return;
+    }
+
+    lastIncomingValue.current = incomingValue;
+    setEditorValue(editor, incomingValue);
+  }, [editor, incomingValue]);
+};
+
+/**
+ * Attaches a custom onChange callback that
+ */
+const useOnValueChanged = (onChange?: (doc: Contentful.Document) => unknown) => {
+  const editor = usePlateSelectors().editor();
+  const setEditorOnChange = usePlateActions().onChange();
+
+  React.useEffect(() => {
+    const cb = createOnChangeCallback(onChange);
+
+    setEditorOnChange({
+      fn: (document) => {
+        // Skip irrelevant events e.g. mouse selection
+        const operations = editor?.operations.filter((op) => {
+          return op.type !== 'set_selection';
+        });
+
+        if (operations.length === 0) {
+          return;
+        }
+
+        cb(document);
+      },
+    });
+  }, [editor, onChange, setEditorOnChange]);
```

Comment:
Does createOnChangeCallback need to be part of the dependencies here?

Reply:
No, this is an external dependency that doesn't change.